### PR TITLE
Add calculate_fee and calculate_fee_rate on wallet

### PR DESCRIPTION
**Pull request by reez**
### Description

~Add calculate_fee on wallet.~ 
Add calculate_fee and calculate_fee_rate on wallet.

Continuing to implement functions/types associated with [PR on bdk "Remove TransactionDetails from Wallet API"](https://github.com/bitcoindevkit/bdk/pull/1048).

### Notes to the reviewers

Open to hearing where to best place the new code in each file too (should `BdkCalculateFeeError` be placed somewhere different in the UDL, etc).

~Currently I have `BdkCalculateFeeError` with associated values, but I am ignoring associated values in the UDL, and will address that in a follow up PR with all of the other changes we need to possibly do to Error types and associating data with them.~ 
I have `CalculateFeeError` with associated values and expose them in the UDL as well.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [ ] I've added docs for the new feature

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [ ] I'm linking the issue being fixed by this PR

### CODE REVIEW

**Review by thunderbiscuit — CHANGES_REQUESTED** [review]
Nice little addition! Only small issue I think is that the name is reversed; you're exposing a new error type called `BdkCalculateFeeError`, but we usually expose the a type with the same name as in Rust and internally transform the Rust type into a `BdkSomethingSomething` type. Just a switcharoo of those will make this PR ready to go.

File: bdk-ffi/src/bdk.udl [hunk 1]
```diff
@@ -82,6 +82,17 @@ enum BdkError {
   "Psbt",
 };
 
+[Error]
+interface CalculateFeeError {
```
Comment (reez) [inline bdk-ffi/src/bdk.udl:86]: using `interface` per [uniffi docs on errors with associated types](https://mozilla.github.io/uniffi-rs/udl/errors.html)

File: bdk-ffi/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        match error {
+            BdkCalculateFeeError::MissingTxOut(out_points) => CalculateFeeError::MissingTxOut {
+                out_points: out_points.iter().map(|op| op.into()).collect(),
+            },
+            BdkCalculateFeeError::NegativeFee(fee) => CalculateFeeError::NegativeFee { fee },
+        }
+    }
+}
+
+impl std::error::Error for CalculateFeeError {}
```
Comment (reez) [inline bdk-ffi/src/error.rs:35]: Implementing the `std::error::Error` trait for `CalculateFeeError` allows error types to integrate with Rust's standard error handling ecosystem

File: bdk-ffi/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+use std::fmt;
+
+#[derive(Debug)]
+pub enum CalculateFeeError {
+    MissingTxOut { out_points: Vec<OutPoint> },
+    NegativeFee { fee: i64 },
+}
+
+impl fmt::Display for CalculateFeeError {
```
Comment (reez) [inline bdk-ffi/src/error.rs:13]: `fmt::Display` implementation for `CalculateFeeError`: 
- `write!` is used to format and append the error message to the `fmt::Formatter`.
- this formatted message is used wherever the display representation of the error is needed

**Comment by reez** [comment]
Updates: 
- did switcharoo of `CalculateFeeError` and `BdkCalculateFeeError` per [Thunder suggestion](https://github.com/bitcoindevkit/bdk-ffi/pull/437#pullrequestreview-1787456514) 
- exposed `CalculateFeeError` with associated values in the UDL per [uniffi](https://mozilla.github.io/uniffi-rs/udl/errors.html) 
- added `calculate_fee_rate`
- added a simplified `FeeRate` constructor and removed the `from_sat_per_vb` constructor 
- updated tests across all 4 platforms    
 

 This is ready for review again, but any suggestions welcome.

File: bdk-jvm/lib/src/test/kotlin/org/bitcoindevkit/LiveWalletTest.kt [hunk 4]
```diff
@@ -55,8 +55,14 @@ class LiveWalletTest {
...
-
         println("Txid is: ${tx.txid()}")
+
+        val txFee = wallet.calculateFee(tx)
+        println("Tx fee is: ${tx.txid()}")
+
+        val feeRate = wallet.calculateFeeRate(tx)
+        println("Tx fee rate is: ${tx.txid()} sat/vB")
+
         esploraClient.broadcast(tx)
```
Comment (thunderbiscuit) [inline bdk-jvm/lib/src/test/kotlin/org/bitcoindevkit/LiveWalletTest.kt:66]: Same comment as above. Small typo.

File: bdk-python/tests/test_live_wallet.py [hunk 5]
```diff
@@ -64,6 +64,11 @@ def test_broadcast_transaction(self):
         walletDidSign = wallet.sign(psbt)
         self.assertTrue(walletDidSign)
         tx = psbt.extract_tx()
+        print(f"Transaction Id: {tx.txid}")
+        fee = wallet.calculate_fee(tx)
+        print(f"Transaction Fee: {fee}")
+        fee_rate = wallet.calculate_fee_rate(tx)
+        print(f"Transaction Fee Rate: {fee_rate} sat/vB")
```
Comment (thunderbiscuit) [inline bdk-python/tests/test_live_wallet.py:71]: I don't think you're printing the value here but just the object. Me thinks it needs `fee_rate.as_sat_per_vb()`.

File: bdk-ffi/src/bdk.udl [hunk 6]
```diff
@@ -82,6 +82,16 @@ enum BdkError {
...
+  MissingTxOut(sequence<OutPoint> out_points);
+  NegativeFee(i64 fee);
```
Comment (thunderbiscuit) [inline bdk-ffi/src/bdk.udl:88]: This is the first time we use this pattern of defining errors as interfaces in the UDL. Can you confirm that the errors print out what we expect them to? Just making sure it all works well.

**Review by thunderbiscuit — CHANGES_REQUESTED** [review]
Awesome work! I have a few comments but overall this is shaping up.

File: bdk-android/lib/src/androidTest/kotlin/org/bitcoindevkit/LiveWalletTest.kt [hunk 7]
```diff
@@ -59,8 +59,14 @@ class LiveWalletTest {
...
         val tx: Transaction = psbt.extractTx()
-
         println("Txid is: ${tx.txid()}")
+
+        val txFee = wallet.calculateFee(tx)
+        println("Tx fee is: ${tx.txid()}")
+
+        val feeRate = wallet.calculateFeeRate(tx)
+        println("Tx fee rate is: ${tx.txid()} sat/vB")
+
```
[the remark this answers is not in the document]
Comment (reez) [reply bdk-android/lib/src/androidTest/kotlin/org/bitcoindevkit/LiveWalletTest.kt]: Definitely, updated-

File: bdk-ffi/src/types.rs [hunk 8]
```diff
@@ -7,8 +7,22 @@ use bdk::KeychainKind;
...
+use bdk::FeeRate as BdkFeeRate;
+
 use std::sync::Arc;
 
+pub struct FeeRate(BdkFeeRate);
+
+impl FeeRate {
+    pub fn new(bdk_fee_rate: BdkFeeRate) -> Self {
+        FeeRate(bdk_fee_rate)
+    }
```
[the remark this answers is not in the document]
Comment (reez) [reply bdk-ffi/src/types.rs]: Indeed, great call, updated-

File: bdk-ffi/src/types.rs [hunk 9]
```diff
@@ -7,8 +7,22 @@ use bdk::KeychainKind;
...
+
+    pub fn as_sat_per_vb(&self) -> f32 {
+        self.0.as_sat_per_vb()
+    }
+}
```
[the remark this answers is not in the document]
Comment (reez) [reply bdk-ffi/src/types.rs:24]: Totally! I wasn't sure exactly which ones I should open up other than `as_sat_per_vb()` so getting your opinion on just adding `sat_per_kwu()` is great. It is added now.

File: bdk-ffi/src/wallet.rs [hunk 10]
```diff
@@ -98,6 +99,19 @@ impl Wallet {
             .map(|tx| Arc::new(tx.tx_node.tx.clone().into()))
             .collect()
     }
+
+    pub fn calculate_fee(&self, tx: &Transaction) -> Result<u64, CalculateFeeError> {
+        self.get_wallet()
+            .calculate_fee(&tx.clone().into())
+            .map_err(|err| err.into())
```
[the remark this answers is not in the document]
Comment (reez) [reply bdk-ffi/src/wallet.rs]: Absolutely, my b, updated-

File: bdk-ffi/src/wallet.rs [hunk 11]
```diff
@@ -473,7 +487,7 @@ impl TxBuilder {
             tx_builder.manually_selected_only();
         }
         if let Some(sat_per_vb) = self.fee_rate {
-            tx_builder.fee_rate(FeeRate::from_sat_per_vb(sat_per_vb));
+            tx_builder.fee_rate(BdkFeeRate::from_sat_per_vb(sat_per_vb));
```
[the remark this answers is not in the document]
Comment (reez) [reply bdk-ffi/src/wallet.rs:490]: Sounds great, added a [new issue](https://github.com/bitcoindevkit/bdk-ffi/issues/438) for it and assigned it to myself.

File: bdk-jvm/lib/src/test/kotlin/org/bitcoindevkit/LiveWalletTest.kt (hunk 4 above)
Comment (reez) [reply bdk-jvm/lib/src/test/kotlin/org/bitcoindevkit/LiveWalletTest.kt]: Definitely, updated-

File: bdk-python/tests/test_live_wallet.py (hunk 5 above)
Comment (reez) [reply bdk-python/tests/test_live_wallet.py]: Definitely, miss by me, updated-

File: bdk-ffi/src/bdk.udl (hunk 6 above)
Comment (reez) [reply bdk-ffi/src/bdk.udl:88]: I just added two tests in the `error.rs` file to address this. Let me know if:
- those two tests make sense to you
- that is the best place to test these two errors

... or any other thoughts you had on it too. Thanks, great suggestion!

File: bdk-ffi/src/wallet.rs [hunk 12]
```diff
@@ -98,6 +99,19 @@ impl Wallet {
             .map(|tx| Arc::new(tx.tx_node.tx.clone().into()))
             .collect()
     }
+
+    pub fn calculate_fee(&self, tx: &Transaction) -> Result<u64, CalculateFeeError> {
+        self.get_wallet()
+            .calculate_fee(&tx.clone().into())
```
[the remark this answers is not in the document]
Comment (reez) [reply bdk-ffi/src/wallet.rs]: Sounds like a plan to me. I added your suggested code snippet:

```
impl From<&Transaction> for BdkTransaction {
    fn from(tx: &Transaction) -> Self { tx.inner.clone() }
}
```

... then updated 4 functions that had a line related to transaction that could have `.clone()` removed.

wallet.rs
- `sent_and_received`
- `calculate_fee`
- `calculate_fee_rate`

esplora.rs
- `broadcast`

... there was still 1 transaction related `.clone()` and it was wallet.rs `transactions` so I added the reference version (we already have the owned version):

```
impl From<&BdkTransaction> for Transaction {
    fn from(tx: &BdkTransaction) -> Self {
        Transaction { inner: tx.clone() }
    }
}
```

... and then updated the `transactions` function code to remove `.clone()`.

I tested `bdk-ffi` `bdk-swift` `bdk-jvm` locally with these changes and tests still passed.

Hopefully that all makes sense to have updated and added along with your suggestion!

**Comment by reez** [comment]
> Awesome work! I have a few comments but overall this is shaping up.

Great suggestions/thoughts/catches. Appreciated!

I updated all of the small items (and just marked the conversation resolved since they had been updated and were now “Outdated”) hopefully that’s cool, but I left the 2 items that I updated code for but you might still want to converse about too: 
- [trait on Transaction](https://github.com/bitcoindevkit/bdk-ffi/pull/437#discussion_r1437850039). I added comments to this as well as updated code, but just leaving it open for any other feedback or thoughts you might have
- [error testing](https://github.com/bitcoindevkit/bdk-ffi/pull/437#discussion_r1437835474). I added comments and code/tests for this but just leaving it open too for any other feedback or thoughts.

**Comment by reez** [comment]
Rebased

**Review by thunderbiscuit — APPROVED** [review]
ACK 3789c1dcd69827247d7a6afaf1bf36506fa311aa. Thanks!

**Pull request merged**